### PR TITLE
daemon: create plugin record when create pipeline

### PR DESCRIPTION
**Pull request by FeelyChau**
Migration test added: https://github.com/alibaba/pipcook/blob/daemon/create-pipeline/.github/workflows/build.yml#L63

### CODE REVIEW

File: packages/sdk/src/utils.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
 // FIXME(feely): move these to core
 export const tunaMirrorURI = 'https://pypi.tuna.tsinghua.edu.cn/simple';
 export const JobStatusValue = [ 'creating', 'running', 'success', 'fail', 'canceled' ];
-export const PluginStatusValue = [ 'installing', 'installed', 'failed', 'pending' ];
+export const PluginStatusValue = [ 'installing', 'installed', 'failed', 'pending', 'init' ];
```
Comment (yorkie) [inline packages/sdk/src/utils.ts:4]:
```suggestion
export const PluginStatusValue = [ 'installing', 'installed', 'failed', 'pending', 'initialized' ];
```

**Comment by yorkie** [comment]
How can we verify the migration scripts are working correctly?

**Comment by FeelyChau** [comment]
> How can we verify the migration scripts are working correctly?

Migration CI is needed.

**Comment by yorkie** [comment]
> > How can we verify the migration scripts are working correctly?
> 
> Migration CI is needed.

Shall we add that check in this PR?

File: packages/daemon/src/app/controller/pipeline.ts [hunk 2]
```diff
@@ -40,8 +40,26 @@ export class PipelineController extends BaseEventController {
...
+    };
+    const plugins = [];
+    await Promise.all(constants.PLUGINS.map(async (pluginType) => {
+      const plugin = await createPlugin(pluginType);
+      if (plugin) {
+        plugins.push(plugin);
+        parsedConfig[`${pluginType}Id`] = plugin?.id;
+        parsedConfig[pluginType] = plugin?.name;
+      }
+    }));
```
Comment (yorkie) [inline packages/daemon/src/app/controller/pipeline.ts:60]: How about returning the plugins via `Promise.all`?

File: packages/daemon/src/app/controller/pipeline.ts [hunk 3]
```diff
@@ -40,8 +40,26 @@ export class PipelineController extends BaseEventController {
     const { name, configUri, config } = this.ctx.request.body;
     const parsedConfig = await parseConfig(configUri || config);
     parsedConfig.name = name;
+    // the plugin name could be git/web url, we need the real name, and create plugin object
+    const createPlugin = async (field: string) => {
+      if (parsedConfig[field]) {
+        const pkg = await this.pluginManager.fetch(parsedConfig[field]);
+        return this.pluginManager.findOrCreateByPkg(pkg);
+      } else {
+        return undefined;
```
Comment (yorkie) [inline packages/daemon/src/app/controller/pipeline.ts:49]: No need to return undefined here.

File: packages/core/src/types/plugins.ts [hunk 4]
```diff
@@ -245,5 +245,9 @@ export enum PluginStatus {
   /**
    * the plugin installation pending
    */
-  PENDING
+  PENDING,
+  /**
+   * the plugin installation not started
+   */
+  INIT
```
Comment (yorkie) [inline packages/core/src/types/plugins.ts:252]: INITIALIZED?

File: packages/core/src/types/database.ts [hunk 5]
```diff
@@ -34,8 +34,9 @@ export enum PipelineStatus {
  * The pipeline database parameters list.
  */
 export type PipelineDBParams = 'dataCollectParams' | 'dataAccessParams' | 'dataProcessParams' | 'modelDefineParams' | 'modelLoadParams' | 'modelTrainParams' | 'modelEvaluateParams';
+export type PipelineDBId = 'dataCollectId' | 'dataAccessId' | 'dataProcessId' | 'modelDefineId' | 'modelLoadId' | 'modelTrainId' | 'modelEvaluateId';
```
Comment (yorkie) [inline packages/core/src/types/database.ts:37]: Is this necessary to specify here? The Id should be the only concept inside daemon?

File: packages/daemon/src/service/plugin.ts [hunk 6]
```diff
@@ -123,7 +126,9 @@ export class PluginManager {
         category: pkg.pipcook.category,
         datatype: pkg.pipcook.datatype,
         dest: pkg.pipcook.target.DESTPATH,
-        status: PluginStatus.INSTALLING
+        from: pkg.pipcook.source.from,
+        uri: pkg.pipcook.source.uri,
```
Comment (yorkie) [inline packages/daemon/src/service/plugin.ts:130]: How about adding a source prefix for `uri` and `from`?

File: packages/daemon/src/app/controller/pipeline.ts [hunk 7]
```diff
@@ -45,19 +45,16 @@ export class PipelineController extends BaseEventController {
...
+    const plugins = (await Promise.all(constants.PLUGINS.map(async (pluginType) => {
       const plugin = await createPlugin(pluginType);
       if (plugin) {
-        plugins.push(plugin);
         parsedConfig[`${pluginType}Id`] = plugin?.id;
         parsedConfig[pluginType] = plugin?.name;
+        return plugin;
       }
-    }));
+    }))).filter((plugin) => plugin);
```
Comment (yorkie) [inline packages/daemon/src/app/controller/pipeline.ts:57]: It's good to exactly define a function `const isNotUndefined = (f) => f !== undefined` for this?

File: packages/core/src/types/database.ts (hunk 5 above)
Comment (yorkie) [reply packages/core/src/types/database.ts]: Forget to remove it?

File: packages/core/src/types/database.ts [hunk 8]
```diff
@@ -34,8 +34,9 @@ export enum PipelineStatus {
...
 
 /**
  * The pipeline database schema.
  */
-export type PipelineDB = Partial<Record<'id' | 'name' | PluginTypeI | PipelineDBParams, string>>;
+export type PipelineDB = Partial<Record<'id' | 'name' | PluginTypeI | PipelineDBParams | PipelineDBId, string>>;
```
Comment (yorkie) [inline packages/core/src/types/database.ts]: Ditto

File: packages/daemon/src/app/controller/pipeline.ts (hunk 7 above)
Comment (yorkie) [reply packages/daemon/src/app/controller/pipeline.ts:57]:
```js
const plugins = (await Promise.all(constants.PLUGINS.map(createPlugin)))
  .filter(isNotUndefined)
  .map((plugin) => {
    parsedConfig[`${pluginType}Id`] = plugin?.id;
    parsedConfig[pluginType] = plugin?.name;
    return plugin
  });

Comment (yorkie) [reply packages/daemon/src/app/controller/pipeline.ts:57]: Otherwise LGTM :)

**Comment by yorkie** [comment]
The last CI prints the following error:

```
  pipeline api.job test
    ✓ prepare
    ✗ create pipeline
      - Error: SQLITE_BUSY: database is locked
    ✗ create job
      - TypeError: Cannot read property 'id' of undefined
    ✗ query job info
      - TypeError: Cannot read property 'id' of undefined
    ✗ list job infos
      - TypeError: Cannot read property 'id' of undefined
    ✗ clean
      - TypeError: Cannot read property 'status' of undefined
```

**Comment by FeelyChau** [comment]
> The last CI prints the following error:
> 
> ```
>   pipeline api.job test
>     ✓ prepare
>     ✗ create pipeline
>       - Error: SQLITE_BUSY: database is locked
>     ✗ create job
>       - TypeError: Cannot read property 'id' of undefined
>     ✗ query job info
>       - TypeError: Cannot read property 'id' of undefined
>     ✗ list job infos
>       - TypeError: Cannot read property 'id' of undefined
>     ✗ clean
>       - TypeError: Cannot read property 'status' of undefined
> ```

Look like a processor still take up the db file.

**Comment by yorkie** [comment]
> https://github.com/alibaba/pipcook/blob/daemon/create-pipeline/.github/workflows/build.yml#L63

How can we verify the db file is corresponding to a specific version?

**Comment by FeelyChau** [comment]
> > https://github.com/alibaba/pipcook/blob/daemon/create-pipeline/.github/workflows/build.yml#L63
> 
> How can we verify the db file is corresponding to a specific version?

No need to do this, this db file is a baseline, all the migration scripts will be tested.

**Comment by yorkie** [comment]
Ok, I suggest moving this file to the repository and renamed like: "baseline-v1.db".

**Comment by FeelyChau** [comment]
> > https://github.com/alibaba/pipcook/blob/daemon/create-pipeline/.github/workflows/build.yml#L63
> 
> How can we verify the db file is corresponding to a specific version?

Maybe we should rename the file to `pipcook-base.db` in OSS to make it more specific.

**Comment by yorkie** [comment]
> > > https://github.com/alibaba/pipcook/blob/daemon/create-pipeline/.github/workflows/build.yml#L63
> > 
> > 
> > How can we verify the db file is corresponding to a specific version?
> 
> Maybe we should rename the file to `pipcook-base.db` in OSS to make it more specific.

1. We use OSS to save the model and dataset because they cost too many spaces really, but the *.db seems not too big that, so IMO it should be saved to daemon/test/fixtures for running the tests without internet.
2. We don't promise the backward-compatible of the database for major upgrades, the filename with version might make sense.

**Comment by FeelyChau** [comment]
> > > > https://github.com/alibaba/pipcook/blob/daemon/create-pipeline/.github/workflows/build.yml#L63
> > > 
> > > 
> > > How can we verify the db file is corresponding to a specific version?
> > 
> > 
> > Maybe we should rename the file to `pipcook-base.db` in OSS to make it more specific.
> 
> 1. We use OSS to save the model and dataset because they cost too many spaces really, but the *.db seems not too big that, so IMO it should be saved to daemon/test/fixtures for running the tests without internet.
> 2. We don't promise the backward-compatible of the database for major upgrades, the filename with version might make sense.

Done.

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -60,6 +60,7 @@ jobs:
         npm test
     - name: daemon start
       run: |
+        mkdir -p ~/.pipcook/db && cp ./packages/daemon/test/res/baseline-1.0.0.db ~/.pipcook/db/pipcook.db
```
Comment (yorkie) [inline .github/workflows/build.yml:63]: .
```suggestion
        mkdir -p ~/.pipcook/db && cp ./packages/daemon/test/res/baseline-v1.db ~/.pipcook/db/pipcook.db
```

Comment (yorkie) [reply .github/workflows/build.yml:63]: Shall we have a restore progress for later steps?

**Review by yorkie — APPROVED** [review]
Looks good to me

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -58,6 +58,9 @@ jobs:
     - name: Running unit tests
       run: |
         npm test
+    - name: migration
+      run: |
+        [ "${{ runner.os }}" -eq "Linux" ] && mkdir -p ~/.pipcook/db && cp ./packages/daemon/test/res/baseline-v1.db ~/.pipcook/db/pipcook.db || true
```
Comment (yorkie) [inline .github/workflows/build.yml:63]: This line is too long :(

**Review by yorkie — APPROVED** [review]

**Review by WenheLI — APPROVED** [review]
LGTM

**Review by rickycao-qy — APPROVED** [review]
LGTM

**Pull request merged**